cip-observe-script-type/README.md

```diff
@@ -0,0 +1,188 @@
+```
+With this pattern DApps are able to process roughly 12-30 forwardNFTValidator UTxO's  per transaction without exceeding script budget limitations.
+The time complexity of this validator is **O(n)** where n is the number of tx inputs. This logic is executed once per input being unlocked  / currency symbol being minted. 
+The redundant execution of searching the inputs for a token is the largest throughput bottleneck for these DApps; it is **O(n*m)** where n is the number of inputs and m is the number of `forwardValidator` inputs + `forwardValidator` minting policies.
+Using the stake validator trick, the time complexity of the forwarding logic is improved to **O(1)**. The forwardValidator logic becomes:
+```haskell
+forwardWithStakeTrick:: StakingCredential -> BuiltinData -> BuiltinData -> ScriptContext -> ()
+forwardWithStakeTrick obsScriptCred tkIdx ctx = fst (head stakeCertPairs) == obsScriptCred 
+  where 
+    info = txInfo ctx 
+    stakeCertPairs = AssocMap.toList (txInfoWdrl info)
+
+stakeValidatorWithSharedLogic :: AssetClass -> BuiltinData -> ScriptContext -> () 
+stakeValidatorWithSharedLogic stateToken _rdmr ctx = assetClassValueOf stateToken (valueSpent (txInfo ctx)) == 1
+```
+For the staking validator trick (demonstrated above), we are simply checking that the StakingCredential of the the staking validator containing the shared validation logic is in the first pair in `txInfoWdrl`. If the StakingCredential is present in `txInfoWdrl`, that means the staking validator (with our shared validation logic) successfully executed in the transaction. This script is **O(1)** in the case where you limit it to one shared logic validator (staking validator), or if you don't want to break composability with other staking validator, 
+then it becomes **O(obs_N)** where `obs_N` is the number of Observe validators that are executed in the transaction as you have to verify that the StakingCredential is present in `txInfoWdrl`.
+
+
+This proposal makes this design pattern indepedent from implementation details of stake validators and withdrawals, and improves efficiency and readability for validators that implement it. 
+
+## Specification
+<!-- The technical specification should describe the proposed improvement in sufficient technical detail. In particular, it should provide enough information that an implementation can be performed solely on the basis of the design in the CIP. This is necessary to facilitate multiple, interoperable implementations. This must include how the CIP should be versioned. If a proposal defines structure of on-chain data it must include a CDDL schema in it's specification.-->
+The type signature of this script type will be consistent with the type signature of minting and staking validators, namely:
+```haskell
+Redeemer -> ScriptContext -> () 
+```
+
+The type signature of the newly introduced `Purpose` will be:
+```haskell
+Observe Integer -- ^ where integer is the index into the observations list.
+```
```

Comment:
Is there a reason why the purpose should include the index instead of the script hash like with minting policies and staking scripts? The spending purpose is the only one that does not provide the hash directly and I only ever use it to lookup the script's hash from the input being spent. It seems other developers do the same since there is now even this [CIP](https://github.com/cardano-foundation/CIPs/pull/754) which suggests changing the spending purpose to include the script's hash. I would much rather have the `Purpose` be:
``` Haskell
Observe ScriptHash -- ^ the hash of the script currently observing.
```

Reply:
I would argue it should be the other way around, the spending purpose should include the index of the UTxO being spent. The standard usage for spending scripts is to find the input that is currently being validated against.

Reply:
> the spending purpose should include the index of the UTxO being spent.

I think this makes sense for spending scripts but not for observation scripts. A `TxInInfo` contains a lot more information than just a `TxOutRef` or a `ScriptHash`. It probably doesn't make sense to put *all* of that information into the `Purpose` so an index place-holder makes sense. But for observation scripts, the only thing you can get from the `TxInfoObservations` is a `ScriptHash`, and a `ScriptHash` fits no problem in the `Purpose`. If the observation index serves no purpose other than to get the script hash for the current observation script, then doesn't it makes more sense to inline the script hash instead of using an index place-holder? Otherwise, the index is forcing the unnecessary extra step of looking up the script hash in the observation list. Do you personally have another use for the observation index in mind?

Reply:
Yes, I agree. 

Spending Script purpose should be `Spending Integer ScriptHash` and observe purpose should be `Observe ScriptHash`

Reply:
This is a weak spot in the current design of things to be sure. The thing that would be closest to the ledger would be to always provide a "pointer" to the thing that caused the run, in which case we should indeed probably change the `Spending` purpose to have an index instead. The current situation is a bit inconsistent; in some cases we resolve some of the information for you, in others not.

Reply:
If we do decide to resolve information for ScriptPurpose, then we should make the information we resolve consistent across all the different constructors. I think `ScriptHash` is a good option in that case, since it is universally useful across the common script types (Minting, Spending, Staking). 

But I really like the idea of integer indices for everything because then we can move towards a future where we can work towards accessing the unique efficiency benefits from transaction determinism that other smart contract platforms (ie EVM) cannot have access to (especially if in the far future we have support for constant access indexed data structures directly in the script context). I feel like currently, we haven't even begun to scratch the surface of utilizing the advantages afforded to us via Cardano's transaction determinism. 

For instance, since we know at the time of transaction construction what the majority of the script context will look like, we can theoretically bring most lookups from O(n) to O(1). Especially with [TxOut Redeemers](https://github.com/cardano-foundation/CIPs/pull/735) you can imagine a world where common lookup operations like `findOwnInput`, `valueOf`, `findStateThreadOutput`, are all O(1). Instead of looking up elements on-chain, we look them up off-chain and provide the index to the on-chain computation which then only has to check that the element is indeed where we claim it is.